lib/zoneinfo_module.c

```diff
@@ -0,0 +1,2194 @@
+
+static PyObject *
+zoneinfo_tzname(PyObject *self, PyObject *dt)
+{
+    _ttinfo *tti = find_ttinfo((PyZoneInfo_ZoneInfo *)self, dt);
+    if (tti == NULL) {
+        return NULL;
+    }
+    Py_INCREF(tti->tzname);
+    return tti->tzname;
+}
+
+#define HASTZINFO(p) (((_PyDateTime_BaseTZInfo *)(p))->hastzinfo)
+#define GET_DT_TZINFO(p) \
+    (HASTZINFO(p) ? ((PyDateTime_DateTime *)(p))->tzinfo : Py_None)
+
+static PyObject *
+zoneinfo_fromutc(PyObject *obj_self, PyObject *dt)
+{
+    if (!PyDateTime_Check(dt)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "fromutc: argument must be a datetime");
+        return NULL;
+    }
+    if (GET_DT_TZINFO(dt) != obj_self) {
+        PyErr_SetString(PyExc_ValueError,
+                        "fromutc: dt.tzinfo "
+                        "is not self");
+        return NULL;
+    }
+
+    PyZoneInfo_ZoneInfo *self = (PyZoneInfo_ZoneInfo *)obj_self;
+
+    int64_t timestamp;
+    if (get_local_timestamp(dt, &timestamp)) {
+        return NULL;
+    }
+    size_t num_trans = self->num_transitions;
+
+    _ttinfo *tti = NULL;
+    unsigned char fold = 0;
+
+    if (num_trans >= 1 && timestamp < self->trans_list_utc[0]) {
+        tti = self->ttinfo_before;
+    }
+    else if (num_trans == 0 ||
+             timestamp > self->trans_list_utc[num_trans - 1]) {
+        tti = find_tzrule_ttinfo_fromutc(&(self->tzrule_after), timestamp,
+                                         PyDateTime_GET_YEAR(dt), &fold);
+
+        // Immediately after the last manual transition, the fold/gap is
+        // between self->trans_ttinfos[num_transitions - 1] and whatever
+        // ttinfo applies immediately after the last transition, not between
+        // the STD and DST rules in the tzrule_after, so we may need to
+        // adjust the fold value.
+        if (num_trans) {
+            _ttinfo *tti_prev = NULL;
+            if (num_trans == 1) {
+                tti_prev = self->ttinfo_before;
+            }
+            else {
+                tti_prev = self->trans_ttinfos[num_trans - 2];
+            }
+            int64_t diff = tti_prev->utcoff_seconds - tti->utcoff_seconds;
+            if (diff > 0 &&
+                timestamp < (self->trans_list_utc[num_trans - 1] + diff)) {
+                fold = 1;
+            }
+        }
+    }
+    else {
+        size_t idx = _bisect(timestamp, self->trans_list_utc, num_trans);
+        _ttinfo *tti_prev = NULL;
+
+        if (idx >= 2) {
+            tti_prev = self->trans_ttinfos[idx - 2];
+            tti = self->trans_ttinfos[idx - 1];
+        }
+        else {
+            tti_prev = self->ttinfo_before;
+            tti = self->trans_ttinfos[0];
+        }
+
+        // Detect fold
+        int64_t shift =
+            (int64_t)(tti_prev->utcoff_seconds - tti->utcoff_seconds);
+        if (shift > (timestamp - self->trans_list_utc[idx - 1])) {
+            fold = 1;
+        }
+    }
+
+    // Mutating this datetime because it's C, lol.
```

Comment:
This comment should probably be moved to line 455 or removed entirely, but AFAIK, there's no chance of this, because `PyNumber_Add` always returns a new instance of `datetime`.

I'll leave this unresolved for the moment to remind me to:

1. move the comment
2. possibly refactor this something like this:

```C
PyObject *tmp = PyNumber_Add(dt, tti->utcoff);
if (tmp != dt) {
    ... // Construct a new datetime with fold=1.
} else if (fold) {
    ((PyDateTime_DateTime *)dt)->fold =1;
}
```
3. Handle the possibility that `PyNumber_Add` will fail - currently it segfaults if you do this:

```python
from datetime import datetime
from zoneinfo._czoneinfo as ZoneInfo

zi = ZoneInfo("America/New_York")
class DT(datetime):
    def __add__(self, other):
        raise Exception("Nope")
```

Oops!

Reply:
Made some tweaks so issue 1 is resolved and issue 3 is fixed but still needs a test.

Another possible approach for issue 2:

```C
PyObject *tmp = PyNumber_Add(dt, tti->utcoff);
if (fold) {
    if (PyDateTime_CheckExact(tmp)) {
        ((PyDateTime_DateTime *)tmp)->fold = 1;
        dt = tmp;
    } else {
        ... // C equivalent of dt = tmp.replace(fold=1)
        if (dt == NULL) {
            Py_DECREF(tmp);
        }
    }
}
```

That should also handle the possibility that subclass + x returns some thing other than a datetime subclass without possible memory corruption.

Reply:
Ok, this has been fixed and a test added.